### PR TITLE
Add a parameter to pass through the initial assembly name

### PR DESCRIPTION
**Pull request by jasonmalinowski**
Right now we don't get the command line string for options until the full design time build has completed; however the project system can still give us the evaluated string which is likely to be close enough. This allows features that want to get symbol names for source to have a better chance of having something that they can use and look up in caches or in the cloud.

This unblocks https://github.com/dotnet/project-system/pull/7216.

### CODE REVIEW

**Review by jmarolf — APPROVED** [review]
I assume the evaluated-model-mode doesn't show diagnostics

**Comment by jasonmalinowski** [comment]
@jmarolf: it shouldn't; right now if we aren't passing this the default assembly name becomes gibberish with a random GUID, which would be even worse for diagnostics!

**Review by ryzngard — APPROVED** [review]

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContextFactory.cs [hunk 1]
```diff
@@ -17,7 +15,11 @@ internal interface IWorkspaceProjectContextFactory
     {
         /// <inheritdoc cref="CreateProjectContextAsync"/>
         [Obsolete("Use CreateProjectContextAsync instead")]
-        IWorkspaceProjectContext CreateProjectContext(string languageName, string projectUniqueName, string projectFilePath, Guid projectGuid, object hierarchy, string binOutputPath);
+        IWorkspaceProjectContext CreateProjectContext(string languageName, string projectUniqueName, string projectFilePath, Guid projectGuid, object? hierarchy, string? binOutputPath);
+
+        /// <inheritdoc cref="CreateProjectContextAsync"/>
+        [Obsolete("Use CreateProjectContextAsync instead")]
+        IWorkspaceProjectContext CreateProjectContext(string languageName, string projectUniqueName, string projectFilePath, Guid projectGuid, object? hierarchy, string? binOutputPath, string? assemblyName);
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContextFactory.cs:22]: 😕

Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContextFactory.cs:22]: So this is "fun": we obsoleted those methods, the project system tried moving to the new methods; that broke, so they rolled back, and we haven't fixed it. @CyrusNajmabadi what's the next step for resolving that?

Comment (drewnoakes) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContextFactory.cs:22]: There was an RPS regression when we tried using the new free-threaded API. The logs for that regression have no expired, so to understand that regression will require another PS insertion.

**Review by drewnoakes — APPROVED** [review]

**Pull request merged**